netcreateorg/netcreate-2018
# Feature: Heartbeat Ping + UADDR-based Database Locking

If a client loses their internet connection while editing a node or edge, the node or edge will remain locked.  The node or edge would be unlocked again when the server is restarted, but that is not ideal when deploying Net.Create in a live classroom environment.

While #107 added a heartbeat emanated by the server so the client could detect when it had lost its internet connection to the server, the server itself could not detect when the client lost its connection.  A disconnected socket will eventually close after about 4 minutes of no activity, but this is far too long for a live classroom.

This pull request introduces two new features:


1. Client + Server Ping

In order for the server to detect when a client connection is lost, the heartbeat system has been revised to be a ping system.  It works like this:

a. As soon as a client connects, the server sends a `ping` message on the websocket.
b. When the client receives the `ping` message, it responds with a `pong` message.
c. The server waits for the `pong` message.  If it doesn't receive one, it takes action and unlocks any nodes or edges locked by the client.


2. UADDR-based Data Locking

In the previous implementation of data locking, locking was done anonymously.  Anyone could lock any object, and we had no idea who did the locking.  In order to release a lock when a client is disconnected, we needed to associate the locks with the client that requested it.  This way, when the server detects a loss of connection, it knows which objects to unlock.

While this is not a fully hardened system, it does improve the overall stability of the system.

This addresses #110.



# To Test

1. Check out `dev-bl/config-ip-ping`.
2. `npm run dev`
3. Client1: Open up `localhost:3000` on Client1 computer.
4. Client1: Select a node (e.g. "tacitus") and click "Edit"
5. Client2: Open up `'localhost:3000' on a second Client2 computer.
6. Client2: Select the same node ("tacitus") and click "Edit" -- you should receive a "locked" message.
7. Client1: Go back to Client1 and disconnect it from the internet, e.g. turn off its wifi.  You should see a "Server Disconnected" message.
8. Wait 5 seconds for the loss of connection to be detected by the server and the record unlocked.
8. Client2: Click the same node and try "Edit" again --  This time you should be able to edit the node because Client1 has released the lock.

Edges should work the same way.
